homeworks/01-bubble-sort/src/App/App.tsx

```diff
@@ -0,0 +1,141 @@
+import SORT_STATUS from '../logic/sortStatus';
+import bubbleSortStep from '../logic/bubbleSort';
+
+// style
+import s from './App.module.scss';
+
+interface AppState {
+  inputValue: number;
+  arr: number[];
+  status: SORT_STATUS;
+}
+
+
+class App extends React.Component<{}, AppState> {
+  private intervalID?: ReturnType<typeof setInterval>;
+  private length: number;
+
+  constructor(props: {}) {
+    super(props);
+    this.length = 10;
+    this.state = {
+      inputValue: this.length,
+      arr: [],
+      status: SORT_STATUS.start
+    }
+  }
+
+  componentDidMount(): void {
+    this.setState({ arr: getRandomArr(this.length) });
+  }
+
+  componentDidUpdate(prevProps: {}, prevState: AppState): void {
+    if (this.state.status !== prevState.status) {
+      switch (this.state.status) {
+        case SORT_STATUS.sorting:
+          this.intervalID = setInterval(this.getSortStep, 100);
+          break;
+        case SORT_STATUS.start:
+        case SORT_STATUS.pause:
+        case SORT_STATUS.finish:
+          this.removeInterval();
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  componentWillUnmount(): void {
+    this.removeInterval();
+  }
+
+  private removeInterval(): void {
+    if (this.intervalID) {
+      clearInterval(this.intervalID)
+    }
+  }
+
+  private handleClickControl: React.MouseEventHandler<HTMLButtonElement> = () => {
+    switch (this.state.status) {
+      case SORT_STATUS.start:
+        this.setState({ status: SORT_STATUS.sorting });
+        break;
+      case SORT_STATUS.sorting:
+        this.setState({ status: SORT_STATUS.pause });
+        break;
+      case SORT_STATUS.pause:
+        this.setState({ status: SORT_STATUS.sorting });
+        break;
+      default:
+        break;
+    }
+  }
+
+  private handleClickReset: React.MouseEventHandler<HTMLButtonElement> = () => {
+    this.setState({ arr: getRandomArr(this.length), status: SORT_STATUS.start });
+  }
+
+  private handleClickManual: React.MouseEventHandler<HTMLButtonElement> = () => {
+    const { status } = this.state;
+
+    if ( status === SORT_STATUS.start || status === SORT_STATUS.pause ) {
+      this.getSortStep();
+    }
+  }
+
+  private handleChangeLength = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    const min = +e.target.min;
+    const max = +e.target.max;
+    const value = parseInt(e.target.value);
+
+    if (value >= min && value <= max) {
+      this.length = value;
+      this.setState({ inputValue: value, arr: getRandomArr(this.length), status: SORT_STATUS.start });
+    } else {
+      this.setState({ inputValue: value });
+    }
+  }
+
+  private getSortStep = (): void => {
+    /*
+      Please, give me some recommendations. Should I use function form of setState below or use simple form setState?
+      I used function form because newArr depended on previous arr state and function getSortStep is used in setInterval.
+     */
+    this.setState(({ arr, status }) => {
```

Comment:
I think both are good. There is no problem if you would do it with setState with object argument.
IMO better to think how to split setState logic and logic of function, for example:
```
private getSortStep = (): void => {
  const { newArr, isSorted } = bubbleSortStep([...this.state.arr]);
  if (isSorted) {
    this.setState({ isSorted: SORT_STATUS.finished })
  } else {
    this.setState({ arr: newArr })
  }
}
```